index.js

```diff
@@ -0,0 +1,130 @@
+import { IPLDURL } from 'js-ipld-url'
+import { create as createTyped } from '@ipld/schema/typed.js'
+import { toDSL } from '@ipld/schema/to-dsl.js'
+import { CID } from 'multiformats/cid'
+import { base32 } from 'multiformats/bases/base32'
+import { base36 } from 'multiformats/bases/base36'
+
+export const DEFAULT_CID_BASES = base32.decoder.or(base36.decoder)
+
+export default class IPLDURLSystem {
+  constructor ({
+    getNode,
+    adls = new Map(),
+    cidBases = DEFAULT_CID_BASES
+  }) {
+    if(!getNode) throw new TypeError('Must provide a getNode function')
+    this.getNode = getNode
+    this.adls = adls
+    this.cidBases = cidBases
+  }
+
+  async resolve (url) {
+    const { hostname: root, segments, searchParams } = new IPLDURL(url)
+
+    const cid = CID.parse(root, this.cidBases).toV1()
+    let data = await this.getNode(cid)
+
+    const initialParameters = {}
+    let shouldProcessRoot = false
+    if (searchParams.has('schema')) {
+      shouldProcessRoot = true
+      initialParameters.schema = searchParams.get('schema')
+      initialParameters.type = searchParams.get('type')
+    }
+    if (searchParams.has('adl')) {
+      // TODO Should other parameters be passed to the ADL function?
+      shouldProcessRoot = true
+      initialParameters.adl = searchParams.get('adl')
+    }
+    if (shouldProcessRoot) {
+      data = await this.#applyParameters(data, initialParameters)
+    }
+
+    // TODO: Account for dag-pb?
+    for (const { name, parameters } of segments) {
+      // This does enables ADLs to return promises for properties
+      data = await data[name]
+      const asCID = CID.asCID(data)
+      if (asCID) {
+        data = await this.getNode(asCID)
+      }
+      data = await this.#applyParameters(data, parameters)
+    }
+
+    return data
+  }
+
+  async #applyParameters (origin, { schema, adl, ...parameters }) {
+    let data = origin
+
+    const asCID = CID.asCID(data)
+    if (asCID) {
+      data = await this.getNode(asCID)
+    }
+
+    if (schema) {
+      data = await SchemaADL(data, { schema, ...parameters }, this)
+    }
+
+    if (adl) {
+      if (!this.adls.has(adl)) {
+        const known = [...this.adls.keys()].join(', ')
+        throw new Error(`Unknown ADL type ${adl}. Must be one of ${known}`)
+      }
+      data = await this.adls.get(adl)(data, parameters, this)
+    }
+
+    return data
+  }
+}
+
+export async function SchemaADL (node, { schema, type }, system) {
+  if (!schema || !type) {
+    throw new TypeError('Must specify which type to use with the schema parameter')
+  }
+
+  const schemaCID = CID.parse(schema, system.cidBases)
+  const schemaDMT = await system.getNode(schemaCID)
+  const converted = makeTyped(node, schemaDMT, type)
```

Comment:
I'm not sure if there's a possibility for caching here, maybe there is since this is a multi-use "system", but if you memoize `converted` then it'll be even more efficient. These are generated functions that cost a little to generate up-front but should be really speedy for multiple uses. `converted` could be cached alongside the schema CID on the parent `IPLDURLSystem`.
Obviously not necessary, but if this gets reused over and over then it might be worth doing at some point.

Reply:
Hmm, Ii'd be worried about having a memory leak. Maybe if there was an LRU? What do you think about having a separate issue to track caching of schemas? It might also be relevant to caching WASM modules once we get those in.

Reply:
also not sure where the caching would take place given Schema is structured as an ADL. 🤔